src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java

```diff
@@ -0,0 +1,201 @@
+
+  private final BlobContainerClient blobContainerClient;
+
+  private final SasTokenGeneratorStrategy sasTokenGeneratorStrategy;
+  private final AzureSasCredential blobContainerSasTokenCreds;
+  private final BlobServiceClient blobServiceClient;
+  private UserDelegationKey delegationKey;
+
+  public BlobContainerClientFactory(String accountName, String accountKey, String containerName) {
+
+    if (StringUtils.isBlank(accountName)) {
+      throw new IllegalArgumentException("The account name is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(accountKey)) {
+      throw new IllegalArgumentException("The account key is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(containerName)) {
+      throw new IllegalArgumentException("The container name is either null or empty.");
+    }
+
+    blobContainerClient =
+        createBlobServiceClientUsingSharedKey(accountName, accountKey)
+            .getBlobContainerClient(containerName);
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.SHARED_KEY;
+    blobContainerSasTokenCreds = null;
+    blobServiceClient = null;
+  }
+
+  public BlobContainerClientFactory(
+      String accountName, TokenCredential azureCredential, String containerName) {
+
+    if (StringUtils.isBlank(accountName)) {
+      throw new IllegalArgumentException("The account name is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(containerName)) {
+      throw new IllegalArgumentException("The container name is either null or empty.");
+    }
+
+    blobServiceClient =
+        createBlobServiceClientUsingTokenCredentials(
+            accountName,
+            Objects.requireNonNull(azureCredential, "Azure token credentials are null."));
+    blobContainerClient = blobServiceClient.getBlobContainerClient(containerName);
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.USER_DELEGATED_KEY;
+    blobContainerSasTokenCreds = null;
+  }
+
+  public BlobContainerClientFactory(String containerURLWithSasToken) {
+
+    BlobUrlParts blobUrl = BlobUrlParts.parse(containerURLWithSasToken);
+
+    blobContainerSasTokenCreds =
+        new AzureSasCredential(blobUrl.getCommonSasQueryParameters().encode());
+    blobContainerClient =
+        new BlobContainerClientBuilder()
+            .httpClient(httpClient)
+            .endpoint(
+                String.format(
+                    Locale.ROOT,
+                    "https://%s/%s",
+                    blobUrl.getHost(),
+                    blobUrl.getBlobContainerName()))
+            .credential(blobContainerSasTokenCreds)
+            .buildClient();
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.CONTAINER_SAS_TOKEN;
+    blobServiceClient = null;
+  }
+
+  public BlobContainerClient getBlobContainerClient() {
+    return blobContainerClient;
+  }
+
+  public String createReadOnlySasUrlForBlob(String blobName) {
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.SHARED_KEY)) {
```

Comment:
Using `if` to compare the same object repeatedly is really `switch`, and using `switch` is usually clearer. Also, since enums are constant values they can be compared using `==` instead of `equals()`.

Finally, another way to express this is by using a method on the enum, which would avoid the need to `switch` and `throw` at the end.

Reply:
Using a method in the enum seems like a pretty good choice here, as the purpose of the enum value here is effectively a dynamic way to create subclass (specializations) of `BlobContainerClientFactory`.

Reply:
Resolving it as we have thread below on this.